resource.go

```diff
@@ -60,13 +82,79 @@ func getResourceField(v interface{}) (f *structs.Field, err error) {
 	return
 }
 
+func getIdFromMap(m map[string]interface{}) (ret int, err error) {
+	if val, ok := m["id"].(int); !ok {
+		ret = 0
+	} else {
+		ret = val
+	}
+	return
+}
+
 // Retrieve Id field from interface.
 func resourceId(v interface{}) (ret int, err error) {
 	var f *structs.Field
+	invalidTypeErr := errors.New("Not a struct, map, or pointer to either")
+	ty := reflect.TypeOf(v)
 
-	if f, err = getResourceField(v); err == nil {
-		ret = f.Value().(int)
+	switch ty.Kind() {
```

Comment:
Why keep `ty` around? also, you can do `switch ty := reflect.TypeOf(v); ty.Kind()` regardless, but I think `switch reflect.TypeOf(v) {` would be nice.

In case you weren't aware, you can just avoid all of this reflection code and do a straight up [type switch](https://golang.org/doc/effective_go.html#type_switch) like:

```go
switch v := v.(type) {
    case int:
        ... deal with int ...
    case string:
        ... deal with string ...   
    ... etc. ...
}
```

But here since you're trying to deal with arbitrary structs + maps I think that wouldn't be appropriate. Mentioning it in case you hadn't seen it before :)